// lost-c3-lib/src/entities/user/Entity.ts
export interface EntityOptions {
    /**
     * A string specifying a unique ID for the ACE. 
     * This is used in the language file. 
     * By convention this is lowercase with dashes for separators, e.g. "my-condition".
     */
    Id: string,
    /**
     * The name that appears in the condition/action picker dialog.
     */
    Name: string,
    /**
     * The text that appears in the event sheet. 
     * You can use simple BBCode tags like [b] and [i], and use {0}, {1} etc. as parameter placeholders. 
     * (There must be one parameter placeholder per parameter.) 
     * For behaviors only, the placeholder {my} is substituted for the behavior name and icon.
     */
    DisplayText: string, 
    /**
     * A description of the action or condition, which appears as a tip at the top of the condition/action picker dialog.
     */
    Description?: string, 

    /**
     * Set to true to highlight the ACE in the condition/action/expression picker dialogs. 
     * This should only be used for the most regularly used ACEs, to help users pick them out from the list easily.
     */
    Highlight?: boolean;
    /**
     * Set to true to deprecate the ACE. 
     * This hides it in the editor, but allows existing projects to continue using it.
     */
    Deprecated?: boolean;
}
