// lost-c3-lib/src/index.ts
/**
 * Config
 */
export { 
    type PluginConfig, type BehaviorConfig, type BehaviorCategory, 
    type PluginCategory, type ExternalRuntimeScript, type ExternalDomScript, 
    type ExternalCSSFile, type CopyToOutputFile, type Config,
    AddonType
} from "./config/Config";

/**
 * Builder
 */
export type { AddonJSON } from "./builder/AddonJSON";
export type { 
    AcesJSON, AceAction, AceCondition, 
    AceExpression, AceParam 
} from "./builder/AcesJSON";
export type { 
    LanguageJSON, LanguageAction, LanguageCondition, 
    LanguageExpression, LanguageParam, LanguagePluginProperty 
} from "./builder/LanguageJSON";

/**
 * Category
 */
export { LostCategory } from "./category/Category";

/**
 * Export user entities
 */
export { Action } from "./entities/user/Action";
export { Condition } from "./entities/user/Condition";
export { Expression } from "./entities/user/Expression";

/**
 * Export entity params
 */
export { 
    NumberParam, StringParam, AnyParam,
    BooleanParam, ComboParam, CmpParam,
    ObjectParam, ObjectNameParam, LayerParam,
    LayoutParam, KeybParam, InstanceVarParam,
    InstanceVarBoolParam, EventVarParam, EventVarBoolParam,
    AnimationParam 
} from "./params/ParamsCollection";
export { 
    ParamType, type ComboItem, type StringParamOptions, 
    type ComboParamOptions, type ObjectParamOptions 
} from "./params/Param";

/**
 * Export plugin properties
 */
export {
    IntegerProperty, FloatProperty, PercentProperty,
    TextProperty, LongTextProperty, CheckProperty,
    FontProperty, ComboProperty, ColorProperty,
    ObjectProperty, GroupProperty
} from "./plugin-properties/PluginPropertiesCollection";

export { 
    type PluginProperty, PluginPropertyType, 
    type ComboPropertyOptions, type ColorPropertyOptions, type ObjectPropertyOptions,
    type IntegerPropertyOptions, type FloatPropertyOptions, type PercentPropertyOptions,
    type TextPropertyOptions, type LongTextPropertyOptions, type CheckPropertyOptions,
    type FontPropertyOptions
} from "./plugin-properties/PluginProperty";
/**
 * Export text formatting functions
 */
export { Bold, Italic, Underline, Strikethrough, Code } from "./misc/lost-text-formatting";
