// lost-c3-lib/dist/plugin-properties/PluginPropertiesCollection.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GroupProperty = exports.ObjectProperty = exports.ColorProperty = exports.ComboProperty = exports.FontProperty = exports.CheckProperty = exports.LongTextProperty = exports.TextProperty = exports.PercentProperty = exports.FloatProperty = exports.IntegerProperty = void 0;
const PluginProperty_1 = require("./PluginProperty");
/**
 * @description
 * an integer number property, always rounded to a whole number.
 */
class IntegerProperty extends PluginProperty_1.PluginProperty {
    constructor(Options) {
        super(Options, PluginProperty_1.PluginPropertyType.INTEGER);
    }
}
exports.IntegerProperty = IntegerProperty;
/**
 * @description
 * a floating-point number property.
 */
class FloatProperty extends PluginProperty_1.PluginProperty {
    constructor(Options) {
        super(Options, PluginProperty_1.PluginPropertyType.FLOAT);
    }
}
exports.FloatProperty = FloatProperty;
/**
 * @description
 * a floating-point number in the range [0-1] represented as a percentage. For example if the user enters 50%, the property will be set to a value of 0.5.
 */
class PercentProperty extends PluginProperty_1.PluginProperty {
    constructor(Options) {
        if (Options.InitialValue > 1)
            Options.InitialValue = 1;
        if (Options.InitialValue < 0)
            Options.InitialValue = 0;
        super(Options, PluginProperty_1.PluginPropertyType.PERCENT);
    }
}
exports.PercentProperty = PercentProperty;
/**
 * @description a field the user can enter a string in to.
 */
class TextProperty extends PluginProperty_1.PluginProperty {
    constructor(Options) {
        super(Options, PluginProperty_1.PluginPropertyType.TEXT);
    }
}
exports.TextProperty = TextProperty;
/**
 * @description the same as "text", but a button with an ellipsis ("...") appears on the right side of the field. The user can click this button to open a dialog to edit a long string more conveniently. This is useful for potentially long content like the project description, or the main text of the Text object.
 */
class LongTextProperty extends PluginProperty_1.PluginProperty {
    constructor(Options) {
        super(Options, PluginProperty_1.PluginPropertyType.LONG_TEXT);
    }
}
exports.LongTextProperty = LongTextProperty;
/**
 * @description a checkbox property, returning a boolean.
 */
class CheckProperty extends PluginProperty_1.PluginProperty {
    constructor(Options) {
        super(Options, PluginProperty_1.PluginPropertyType.CHECK);
    }
}
exports.CheckProperty = CheckProperty;
/**
 * @description a field which displays the name of a font and provides a button to open a font picker dialog. The property is set to a string of the name of the font.
 */
class FontProperty extends PluginProperty_1.PluginProperty {
    constructor(Options) {
        super(Options, PluginProperty_1.PluginPropertyType.FONT);
    }
}
exports.FontProperty = FontProperty;
/**
 * @description a dropdown list property. The property is set to the zero-based index of the chosen item. The items field of the options object must be used to specify the available items.
 */
class ComboProperty extends PluginProperty_1.PluginProperty {
    constructor(Options) {
        super(Options, PluginProperty_1.PluginPropertyType.COMBO);
    }
}
exports.ComboProperty = ComboProperty;
/**
 * @description For plugins only — a color picker property. The initial value must be an array, e.g. [1, 0, 0] for red.
 */
class ColorProperty extends PluginProperty_1.PluginProperty {
    constructor(Options) {
        super(Options, PluginProperty_1.PluginPropertyType.COLOR);
    }
}
exports.ColorProperty = ColorProperty;
/**
 * @description For plugins only — an object picker property allowing the user to pick an object class. Note: At runtime, this passes a SID (Serialization ID) for the chosen object class, or -1 if none was picked. Use the runtime method GetObjectClassBySID to look up the corresponding ObjectClass.
 */
class ObjectProperty extends PluginProperty_1.PluginProperty {
    constructor(Options) {
        super(Options, PluginProperty_1.PluginPropertyType.OBJECT);
    }
}
exports.ObjectProperty = ObjectProperty;
/**
 * @description creates a new group in the Properties Bar. There is no value associated with this property.
 */
class GroupProperty extends PluginProperty_1.PluginProperty {
    constructor(Options) {
        super(Options, PluginProperty_1.PluginPropertyType.GROUP);
    }
}
exports.GroupProperty = GroupProperty;
