// lost-c3-lib/dist/plugin-properties/PluginProperty.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PluginProperty = exports.PluginPropertyType = void 0;
var PluginPropertyType;
(function (PluginPropertyType) {
    PluginPropertyType["INTEGER"] = "integer";
    PluginPropertyType["FLOAT"] = "float";
    PluginPropertyType["PERCENT"] = "percent";
    PluginPropertyType["TEXT"] = "text";
    PluginPropertyType["LONG_TEXT"] = "longtext";
    PluginPropertyType["CHECK"] = "check";
    PluginPropertyType["FONT"] = "font";
    PluginPropertyType["COMBO"] = "combo";
    PluginPropertyType["COLOR"] = "color";
    PluginPropertyType["OBJECT"] = "object";
    PluginPropertyType["GROUP"] = "group"; /**
    creates a new group in the Properties Bar. There is no value associated with this property.
    */
})(PluginPropertyType || (exports.PluginPropertyType = PluginPropertyType = {}));
class PluginProperty {
    constructor(Options, Type) {
        this.Options = Options;
        this.Type = Type;
    }
}
exports.PluginProperty = PluginProperty;
