// lost-c3-lib/dist/category/Category.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LostCategory = void 0;
class LostCategory {
    constructor(Options) {
        this.Deprecated = false;
        this.InDevelopment = false;
        this.Id = Options.Id;
        this.Name = Options.Name;
        this.Deprecated = Options.Deprecated || this.Deprecated;
        this.InDevelopment = Options.InDevelopment || this.InDevelopment;
        this.Actions = [];
        this.Conditions = [];
        this.Expressions = [];
        const prototypeActions = Object.getPrototypeOf(this).Actions;
        if (prototypeActions) {
            this.Actions = prototypeActions;
        }
        const prototypeConditions = Object.getPrototypeOf(this).Conditions;
        if (prototypeConditions) {
            this.Conditions = prototypeConditions;
        }
        const prototypeExpressions = Object.getPrototypeOf(this).Expressions;
        if (prototypeExpressions) {
            this.Expressions = prototypeExpressions;
        }
    }
}
exports.LostCategory = LostCategory;
