// lost-c3-lib/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Code = exports.Strikethrough = exports.Underline = exports.Italic = exports.Bold = exports.PluginPropertyType = exports.GroupProperty = exports.ObjectProperty = exports.ColorProperty = exports.ComboProperty = exports.FontProperty = exports.CheckProperty = exports.LongTextProperty = exports.TextProperty = exports.PercentProperty = exports.FloatProperty = exports.IntegerProperty = exports.AnimationParam = exports.EventVarBoolParam = exports.EventVarParam = exports.InstanceVarBoolParam = exports.InstanceVarParam = exports.KeybParam = exports.LayoutParam = exports.LayerParam = exports.ObjectNameParam = exports.ObjectParam = exports.CmpParam = exports.ComboParam = exports.BooleanParam = exports.AnyParam = exports.StringParam = exports.NumberParam = exports.Expression = exports.Condition = exports.Action = exports.LostCategory = exports.AddonType = void 0;
/**
 * Config
 */
var Config_1 = require("./config/Config");
Object.defineProperty(exports, "AddonType", { enumerable: true, get: function () { return Config_1.AddonType; } });
/**
 * Category
 */
var Category_1 = require("./category/Category");
Object.defineProperty(exports, "LostCategory", { enumerable: true, get: function () { return Category_1.LostCategory; } });
/**
 * Export user entities
 */
var Action_1 = require("./entities/user/Action");
Object.defineProperty(exports, "Action", { enumerable: true, get: function () { return Action_1.Action; } });
var Condition_1 = require("./entities/user/Condition");
Object.defineProperty(exports, "Condition", { enumerable: true, get: function () { return Condition_1.Condition; } });
var Expression_1 = require("./entities/user/Expression");
Object.defineProperty(exports, "Expression", { enumerable: true, get: function () { return Expression_1.Expression; } });
/**
 * Export entity params
 */
var ParamsCollection_1 = require("./params/ParamsCollection");
Object.defineProperty(exports, "NumberParam", { enumerable: true, get: function () { return ParamsCollection_1.NumberParam; } });
Object.defineProperty(exports, "StringParam", { enumerable: true, get: function () { return ParamsCollection_1.StringParam; } });
Object.defineProperty(exports, "AnyParam", { enumerable: true, get: function () { return ParamsCollection_1.AnyParam; } });
Object.defineProperty(exports, "BooleanParam", { enumerable: true, get: function () { return ParamsCollection_1.BooleanParam; } });
Object.defineProperty(exports, "ComboParam", { enumerable: true, get: function () { return ParamsCollection_1.ComboParam; } });
Object.defineProperty(exports, "CmpParam", { enumerable: true, get: function () { return ParamsCollection_1.CmpParam; } });
Object.defineProperty(exports, "ObjectParam", { enumerable: true, get: function () { return ParamsCollection_1.ObjectParam; } });
Object.defineProperty(exports, "ObjectNameParam", { enumerable: true, get: function () { return ParamsCollection_1.ObjectNameParam; } });
Object.defineProperty(exports, "LayerParam", { enumerable: true, get: function () { return ParamsCollection_1.LayerParam; } });
Object.defineProperty(exports, "LayoutParam", { enumerable: true, get: function () { return ParamsCollection_1.LayoutParam; } });
Object.defineProperty(exports, "KeybParam", { enumerable: true, get: function () { return ParamsCollection_1.KeybParam; } });
Object.defineProperty(exports, "InstanceVarParam", { enumerable: true, get: function () { return ParamsCollection_1.InstanceVarParam; } });
Object.defineProperty(exports, "InstanceVarBoolParam", { enumerable: true, get: function () { return ParamsCollection_1.InstanceVarBoolParam; } });
Object.defineProperty(exports, "EventVarParam", { enumerable: true, get: function () { return ParamsCollection_1.EventVarParam; } });
Object.defineProperty(exports, "EventVarBoolParam", { enumerable: true, get: function () { return ParamsCollection_1.EventVarBoolParam; } });
Object.defineProperty(exports, "AnimationParam", { enumerable: true, get: function () { return ParamsCollection_1.AnimationParam; } });
/**
 * Export plugin properties
 */
var PluginPropertiesCollection_1 = require("./plugin-properties/PluginPropertiesCollection");
Object.defineProperty(exports, "IntegerProperty", { enumerable: true, get: function () { return PluginPropertiesCollection_1.IntegerProperty; } });
Object.defineProperty(exports, "FloatProperty", { enumerable: true, get: function () { return PluginPropertiesCollection_1.FloatProperty; } });
Object.defineProperty(exports, "PercentProperty", { enumerable: true, get: function () { return PluginPropertiesCollection_1.PercentProperty; } });
Object.defineProperty(exports, "TextProperty", { enumerable: true, get: function () { return PluginPropertiesCollection_1.TextProperty; } });
Object.defineProperty(exports, "LongTextProperty", { enumerable: true, get: function () { return PluginPropertiesCollection_1.LongTextProperty; } });
Object.defineProperty(exports, "CheckProperty", { enumerable: true, get: function () { return PluginPropertiesCollection_1.CheckProperty; } });
Object.defineProperty(exports, "FontProperty", { enumerable: true, get: function () { return PluginPropertiesCollection_1.FontProperty; } });
Object.defineProperty(exports, "ComboProperty", { enumerable: true, get: function () { return PluginPropertiesCollection_1.ComboProperty; } });
Object.defineProperty(exports, "ColorProperty", { enumerable: true, get: function () { return PluginPropertiesCollection_1.ColorProperty; } });
Object.defineProperty(exports, "ObjectProperty", { enumerable: true, get: function () { return PluginPropertiesCollection_1.ObjectProperty; } });
Object.defineProperty(exports, "GroupProperty", { enumerable: true, get: function () { return PluginPropertiesCollection_1.GroupProperty; } });
var PluginProperty_1 = require("./plugin-properties/PluginProperty");
Object.defineProperty(exports, "PluginPropertyType", { enumerable: true, get: function () { return PluginProperty_1.PluginPropertyType; } });
/**
 * Export text formatting functions
 */
var lost_text_formatting_1 = require("./misc/lost-text-formatting");
Object.defineProperty(exports, "Bold", { enumerable: true, get: function () { return lost_text_formatting_1.Bold; } });
Object.defineProperty(exports, "Italic", { enumerable: true, get: function () { return lost_text_formatting_1.Italic; } });
Object.defineProperty(exports, "Underline", { enumerable: true, get: function () { return lost_text_formatting_1.Underline; } });
Object.defineProperty(exports, "Strikethrough", { enumerable: true, get: function () { return lost_text_formatting_1.Strikethrough; } });
Object.defineProperty(exports, "Code", { enumerable: true, get: function () { return lost_text_formatting_1.Code; } });
