// lost-c3-lib/dist/params/ParamsCollection.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AnimationParam = exports.EventVarBoolParam = exports.EventVarParam = exports.InstanceVarBoolParam = exports.InstanceVarParam = exports.KeybParam = exports.LayoutParam = exports.LayerParam = exports.ObjectNameParam = exports.ObjectParam = exports.CmpParam = exports.ComboParam = exports.BooleanParam = exports.AnyParam = exports.StringParam = exports.NumberParam = void 0;
const Param_1 = require("./Param");
/**
* @description A number parameter
* @functiontype number
*/
class NumberParam extends Param_1.Param {
    constructor(Options) {
        super(Options, "number" /* ParamType.NUMBER */);
    }
}
exports.NumberParam = NumberParam;
/**
 * @description A string parameter
 * @functiontype string
 */
class StringParam extends Param_1.Param {
    constructor(Options) {
        super(Options, "string" /* ParamType.STRING */);
    }
}
exports.StringParam = StringParam;
/**
 * @description Either a number or a string
 * @functiontype any
 */
class AnyParam extends Param_1.Param {
    constructor(Options) {
        super(Options, "any" /* ParamType.ANY */);
    }
}
exports.AnyParam = AnyParam;
/**
 * @description A boolean parameter, displayed as a checkbox
 * @functiontype boolean
 */
class BooleanParam extends Param_1.Param {
    constructor(Options) {
        super(Options, "boolean" /* ParamType.BOOLEAN */);
    }
}
exports.BooleanParam = BooleanParam;
/**
 * @description A dropdown list. Items must be specified with the "Items" property.
 * @functiontype number (index of chosen item)
 */
class ComboParam extends Param_1.Param {
    constructor(Options) {
        super(Options, "combo" /* ParamType.COMBO */);
    }
}
exports.ComboParam = ComboParam;
/**
 * @description A dropdown list with comparison options like "equal to", "less than" etc.
 */
class CmpParam extends Param_1.Param {
    constructor(Options) {
        super(Options, "cmp" /* ParamType.CMP */);
    }
}
exports.CmpParam = CmpParam;
/**
 * @description An object picker. The types of plugin to show can be filtered using an optional "allowedPluginIds" property.
 */
class ObjectParam extends Param_1.Param {
    constructor(Options) {
        super(Options, "object" /* ParamType.OBJECT */);
    }
}
exports.ObjectParam = ObjectParam;
/**
 * @description A string parameter which is interpreted as an object name
 */
class ObjectNameParam extends Param_1.Param {
    constructor(Options) {
        super(Options, "objectname" /* ParamType.OBJECT_NAME */);
    }
}
exports.ObjectNameParam = ObjectNameParam;
/**
 * @description A string parameter which is interpreted as a layer name
 */
class LayerParam extends Param_1.Param {
    constructor(Options) {
        super(Options, "layer" /* ParamType.LAYER */);
    }
}
exports.LayerParam = LayerParam;
/**
 * @description A dropdown list with every layout in the project
 */
class LayoutParam extends Param_1.Param {
    constructor(Options) {
        super(Options, "layout" /* ParamType.LAYOUT */);
    }
}
exports.LayoutParam = LayoutParam;
/**
 * @description A keyboard key picker
 */
class KeybParam extends Param_1.Param {
    constructor(Options) {
        super(Options, "keyb" /* ParamType.KEYB */);
    }
}
exports.KeybParam = KeybParam;
/**
 * @description A dropdown list with the non-boolean instance variables the object has
 */
class InstanceVarParam extends Param_1.Param {
    constructor(Options) {
        super(Options, "instancevar" /* ParamType.INSTANCE_VAR */);
    }
}
exports.InstanceVarParam = InstanceVarParam;
/**
 * @description A dropdown list with the boolean instance variables the object has
 */
class InstanceVarBoolParam extends Param_1.Param {
    constructor(Options) {
        super(Options, "instancevarbool" /* ParamType.INSTANCE_VAR_BOOL */);
    }
}
exports.InstanceVarBoolParam = InstanceVarBoolParam;
/**
 * @description A dropdown list with non-boolean event variables in scope
 */
class EventVarParam extends Param_1.Param {
    constructor(Options) {
        super(Options, "eventvar" /* ParamType.EVENT_VAR */);
    }
}
exports.EventVarParam = EventVarParam;
/**
 * @description A dropdown list with boolean event variables in scope
 */
class EventVarBoolParam extends Param_1.Param {
    constructor(Options) {
        super(Options, "eventvarbool" /* ParamType.EVENT_VAR_BOOL */);
    }
}
exports.EventVarBoolParam = EventVarBoolParam;
/**
 * @description A string parameter which is interpreted as an animation name in the object
 */
class AnimationParam extends Param_1.Param {
    constructor(Options) {
        super(Options, "animation" /* ParamType.ANIMATION */);
    }
}
exports.AnimationParam = AnimationParam;
